fairscale/experimental/nn/auto_shard.py

```diff
@@ -0,0 +1,188 @@
+                    node_name_to_shard_id[node_name] = min_shard_id
+                    if node_name == min_node_name:
+                        break
+                shard_id = min_shard_id
+                # TODO(anj-s): Find a way to raise an error early if this can cause OOM errors.
+                shard_to_param_count = _create_shard_to_param_count(param_count, node_name_to_shard_id)
+
+            # Update state that is tracking node -> shard id and shard id -> param count.
+            node_name_to_shard_id[node.name] = shard_id
+            nodes_so_far.append(node.name)
+            # TODO(anj): This could just be an update, we don't need to recreate the map.
+            shard_to_param_count = _create_shard_to_param_count(param_count, node_name_to_shard_id)
+            # If we have gone over the number of params per shard count that we want to
+            # achieve, we should add a new shard.
+            # The shard_id may not have been updated in the map if we are at a node that does not
+            # have params.
+            if shard_id in shard_to_param_count and shard_to_param_count[shard_id] > per_shard_param:
+                shard_id += 1
+        elif node.op == "output":
+            break
+    return node_name_to_shard_id
+
+
+def shard_model(model: torch.nn.Module, shard_count: int = 3) -> List[torch.fx.GraphModule]:
+    """Utility used to shard a model using torch.fx.
+
+    This function traces the model twice in an attempt to identify the
+    right cutpoints and then shard the model. In the first pass we calculate
+    the number of parameters as we are tracing the graph and mark nodes at 
+    which we might want to create a new module. In the second pass we 
+    modify the graph by inserting placeholders and output nodes to essentially
+    shard the graph.
+
+    We don't support skip connections between shards. This means that all 
+    input and output is self contained within a given shard. A node from
+    shard 1 cannot be an input to a node from shard 3. We expect all inputs
+    to a given shard to be coming from the last node in the previous shard.
+    This means that we may not be able to shard models by the specified
+    `shard_count` mentioned by the user. 
+
+    Args:
+        model (nn.Module): Model to be sharded as specified by the device count.
+
+        shard_count (int): Number of shards that we want to split the model into.
+
+    """
+    module_list: List[torch.fx.GraphModule] = []
+    num_graphs = 0
+    new_graph = torch.fx.Graph()  # type: ignore
+    env: Dict[str, Node] = {}
+    new_input_node = None
+    # This is the first pass where we attempt to get a map of where
+    # we need to insert placeholder and output nodes.
+    node_name_to_shard_id = _split_nodes(model, shard_count=shard_count)
+
+    traced_graph_module = torch.fx.symbolic_trace(model)
+
+    # dummy value which indicates that this is the first node.
+    prev_shard_id = 1000
```

Comment:
Nit: why not use an obvious invalid value like `-1`?

Reply:
prev_shard_id needs to be a higher value than the first shard id which is why I picked this.